core/src/nodes/collection.rs

```diff
@@ -110,6 +109,143 @@ pub enum CollectionEvent<TOutEvent> {
+        ReachAttemptId(self.id)
+    }
+
+    /// Returns `true` if accepting this reached node would replace an existing connection to that
+    /// node.
+    #[inline]
+    pub fn would_replace(&self) -> bool {
+        self.parent.nodes.contains_key(&self.peer_id)
+    }
+
+    /// Accepts the new node.
+    pub fn accept(self) -> (CollectionNodeAccept, PeerId) {
+        // Set the state of the task to `Connected`.
+        let former_task_id = self.parent.nodes.insert(self.peer_id.clone(), self.id);
+        let _former_state = self.parent.tasks.insert(self.id, TaskState::Connected(self.peer_id.clone()));
+        debug_assert_eq!(_former_state, Some(TaskState::Pending));
+
+        // It is possible that we already have a task connected to the same peer. In this
+        // case, we need to emit a `NodeReplaced` event.
+        let ret_value = if let Some(former_task_id) = former_task_id {
+            self.parent.inner.task(former_task_id)
+                .expect("whenever we receive a TaskClosed event or close a node, we remove the \
+                         corresponding entry from self.nodes ; therefore all elements in \
+                         self.nodes are valid tasks in the HandledNodesTasks ; qed")
+                .close();
+            let _former_other_state = self.parent.tasks.remove(&former_task_id);
+            debug_assert_eq!(_former_other_state, Some(TaskState::Connected(self.peer_id.clone())));
+
+            // TODO: we unfortunately have to clone the peer id here
+            (CollectionNodeAccept::ReplacedExisting, self.peer_id.clone())
+        } else {
+            // TODO: we unfortunately have to clone the peer id here
+            (CollectionNodeAccept::NewEntry, self.peer_id.clone())
+        };
+
+        // Don't run the destructor.
+        mem::forget(self);
+
+        ret_value
+    }
+
+    /// Denies the node.
+    ///
+    /// Has the same effect as dropping the event without accepting it.
+    #[inline]
+    pub fn deny(self) -> PeerId {
+        // TODO: we unfortunately have to clone the id here, in order to be explicit
+        let peer_id = self.peer_id.clone();
+        drop(self);  // Just to be explicit
+        peer_id
+    }
+}
+
+impl<'a, TInEvent, TOutEvent> fmt::Debug for CollectionReachEvent<'a, TInEvent, TOutEvent> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        f.debug_struct("CollectionReachEvent")
+            .field("peer_id", &self.peer_id)
+            .field("reach_attempt_id", &self.reach_attempt_id())
+            .finish()
+    }
+}
+
+impl<'a, TInEvent, TOutEvent> Drop for CollectionReachEvent<'a, TInEvent, TOutEvent> {
+    fn drop(&mut self) {
+        let task_state = self.parent.tasks.remove(&self.id);
+        debug_assert!(if let Some(TaskState::Pending) = task_state { true } else { false });
+        self.parent.inner.task(self.id).unwrap().close();       // TODO: prove the unwrap
```

Comment:
Would be nice to turn this `unwrap` into an `expect` explaining why it's safe to do so here.